kernel: Init the base.slice_ticks for dummy_thread

hunterhu3000 opened the pull request:
Fixes #46324
Set dummy_thread->base.slice_ticks to 0 when
CONFIG_TIMESLICE_PER_THREAD is set. To avoid
_current_cpu->slice_ticks be a big number.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

andyross reviewed (changes requested):
If I'm reading the linked bug correctly, you have the kernel failing to return from a k_sleep() becuase of an uninitialized (which presumably means zero, right?) slice_ticks field?

That sounds like a kernel bug, not a test thing.  A thread with a zero slice_ticks simply shouldn't be timeslicing, it **absolutely** shouldn't be failing to sleep correctly.  There are tests in the tree that exercise that, so it "should" work.  But clearly it isn't for you?

Do you have an exerciser on any of the in-tree emulation platforms?  We need to figure this out and find the root cause.

RuibinChang commented:
`z_sched_init()` ->` k_sched_time_slice_set(0 , 0)` -> `z_reset_time_slice()` -> `slice_time(curr)`

`curr` is dummy thread and doesn't init, so `curr->base.slice_ticks` is a big number.

Then in `_current_cpu->slice_ticks = slice_time(curr) + sys_clock_elapsed()` of `z_reset_time_slice()`, so `_current_cpu->slice_ticks` is a bigger number, too. This causes the issue.

How about that we add 
```
#ifdef CONFIG_TIMESLICE_PER_THREAD
	dummy_thread->base.slice_ticks = 0;
#endif
```
 in `z_dummy_thread_init()` to init `curr->base.slice_ticks`?

hunterhu3000 commented:
> If I'm reading the linked bug correctly, you have the kernel failing to return from a k_sleep() becuase of an uninitialized (which presumably means zero, right?) slice_ticks field?
> 
> That sounds like a kernel bug, not a test thing. A thread with a zero slice_ticks simply shouldn't be timeslicing, it **absolutely** shouldn't be failing to sleep correctly. There are tests in the tree that exercise that, so it "should" work. But clearly it isn't for you?
> 
> Do you have an exerciser on any of the in-tree emulation platforms? We need to figure this out and find the root cause.

The uninitialized value is not zero, but a magic number. Print the value at the beginning of the test, you will see the log:
```
START - test_priority_cooperative
@@@ _test_priority_cooperativ: before k_thread_create current_cpu->slice_ticks = -1431655766
```
For MEC17, it also has the problem, but the magic number is smaller, about 14000, so MEC can pass the test. Do you think we shoud set current_cpu->slice_ticks = 0 in  k_thread_create when CONFIG_TIMESLICE_PER_THREAD=y ?

hunterhu3000 commented:
> `z_sched_init()` ->` k_sched_time_slice_set(0 , 0)` -> `z_reset_time_slice()` -> `slice_time(curr)`
> 
> `curr` is dummy thread and doesn't init, so `curr->base.slice_ticks` is a big number.
> 
> Then returns to `z_reset_time_slice()`, `_current_cpu->slice_ticks = slice_time(curr) + sys_clock_elapsed()`, so `_current_cpu->slice_ticks` is a bigger number, too. This causes the issue.
> 
> How about that we add
> 
> ```
> #ifdef CONFIG_TIMESLICE_PER_THREAD
> 	dummy_thread->base.slice_ticks = 0;
> #endif
> ```
> 
> in `z_dummy_thread_init()` to init `curr->base.slice_ticks`?

Looks good. And verified on ITE.

hunterhu3000 committed:
kernel: Init the base.slice_ticks for dummy_thread

Fixes #46324
Set dummy_thread->base.slice_ticks to 0 when
CONFIG_TIMESLICE_PER_THREAD is set. To avoid
_current_cpu->slice_ticks be a big number.

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

hunterhu3000 commented:
Take RuibinChang's suggestion to modify in z_dummy_thread_init. Thanks to review

andyross reviewed (approved):
Ah... OK.  This now makes sense to me.  Thanks so much for your attention here.

Interestingly this is not the first time we've been burned by the "thread struct initialization is done piecewise so some fields are uninitialized" problem.   There has historically been resistance to initializing threads with e.g. a memset/struct-assignment, as some architectures put big data structures in there.  But it's really fragile, and maybe it's time to fix this?

hunterhu3000 commented:
It seems some CI test fail.
Some tests of 
···
twister -T tests/kernel/sched/schedule_api/
···
on qemu also fail
